<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="description" content="Bayesembler : Bayesian transcriptome assembly">

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>Bayesembler</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/bioinformatics-centre/bayesembler">View on GitHub</a>

          <h1 id="project_title">Bayesembler</h1>
          <h2 id="project_tagline">Bayesian transcriptome assembly</h2>

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/bioinformatics-centre/bayesembler/zipball/master">Download this project as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/bioinformatics-centre/bayesembler/tarball/master">Download this project as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">
        <p>The Bayesembler is a Bayesian method for doing transcriptome assembly from RNA-seq data.</p>

<h2>
<a id="news" class="anchor" href="#news" aria-hidden="true"><span class="octicon octicon-link"></span></a>News</h2>

<ul>
<li>
<strong>February 2, 2015:</strong> Bayesembler v1.2.0 is released (<a href="https://github.com/bioinformatics-centre/bayesembler/releases/tag/v1.2.0">release notes</a>)</li>
<li>
<strong>November 3, 2014:</strong> The Bayesembler <a href="http://genomebiology.com/2014/15/10/501/abstract">paper</a> is out in Genome Biology.</li>
<li>
<strong>September 4, 2014:</strong> The Bayesembler manuscript has been accepted for publication.</li>
</ul>

<h2>
<a id="in-the-making" class="anchor" href="#in-the-making" aria-hidden="true"><span class="octicon octicon-link"></span></a>In the making</h2>

<p>We are currently working on implementing the following features:</p>

<ul>
<li>New splice-graph engine (we currently rely on a third party engine) that will support combining replicates for splice-graph assembly</li>
<li>Single-end read support</li>
<li>Support for other mappers (e.g. STAR)</li>
<li>Improved support for unstranded data (assembly accuracy on unstranded data is currently not optimal)</li>
<li>Improved parallelisation efficiency</li>
</ul>

<p>Watch us on <strong>GitHub</strong> or follow us on <strong>Twitter (@bayesembler)</strong> to stay informed. Please use the <a href="https://github.com/bioinformatics-centre/bayesembler/issues">GitHub issue tracking system</a> to request any features you would like us to implement.</p>

<h2>
<a id="installing-the-bayesembler" class="anchor" href="#installing-the-bayesembler" aria-hidden="true"><span class="octicon octicon-link"></span></a>Installing the Bayesembler</h2>

<p>The Bayesembler runs on both Linux and OS X and is freely available under the <a href="http://opensource.org/licenses/MIT">MIT License</a>. Static Linux x86_64 builds are available under releases (<a href="https://github.com/bioinformatics-centre/bayesembler/releases/latest">latest release</a>). Please check out the <a href="https://github.com/bioinformatics-centre/bayesembler/wiki/Installation">installation-wiki</a> for further information including how to build the Bayesembler from source.</p>

<h2>
<a id="running-the-bayesembler" class="anchor" href="#running-the-bayesembler" aria-hidden="true"><span class="octicon octicon-link"></span></a>Running the Bayesembler</h2>

<h3>
<a id="input" class="anchor" href="#input" aria-hidden="true"><span class="octicon octicon-link"></span></a>Input</h3>

<p>The Bayesembler requires <em>paired-end</em> RNA-seq data mapped using <a href="http://ccb.jhu.edu/software/tophat/index.shtml">TopHat2</a>.</p>

<h3>
<a id="assembly" class="anchor" href="#assembly" aria-hidden="true"><span class="octicon octicon-link"></span></a>Assembly</h3>

<p>The Bayesembler can be invoked with default settings using</p>

<pre><code>bayesembler -b &lt;tophat2_map.bam&gt;
</code></pre>

<p>Important options include:</p>

<ul>
<li>
<code>-s</code>: Data is strand-specific. Use <code>-s first</code> to indicate mate orientation as in the dUTP protocol or <code>-s second</code> to indicate opposite orientation. If not provided, the program assumes unstranded data (beware that performance on unstranded data is currently not optimal, we are working on improving this).</li>
<li>
<code>-p</code>: Number of threads used for assembly.</li>
</ul>

<h3>
<a id="output" class="anchor" href="#output" aria-hidden="true"><span class="octicon octicon-link"></span></a>Output</h3>

<p>Assembly in <a href="http://genome.ucsc.edu/FAQ/FAQformat.html#format4">Gene Annotation Format</a> (GTF) as <code>assembly.gtf</code>.</p>

<p>Please consult the <a href="https://github.com/bioinformatics-centre/bayesembler/wiki/Manual">manual-wiki</a> for additional options and information about the output.</p>

<h2>
<a id="bug-reports-help-and-feature-requests" class="anchor" href="#bug-reports-help-and-feature-requests" aria-hidden="true"><span class="octicon octicon-link"></span></a>Bug reports, help and feature requests</h2>

<p>We want to know if you are experiencing problems with our program or really need us to implement a new feature. To get in touch with us, please use the <a href="https://github.com/bioinformatics-centre/bayesembler/issues">GitHub issue tracking system</a> or <a href="mailto:bayesembler@binf.ku.dk">bayesembler@binf.ku.dk</a>.</p>

<h2>
<a id="citation-and-credits" class="anchor" href="#citation-and-credits" aria-hidden="true"><span class="octicon octicon-link"></span></a>Citation and credits</h2>

<p>The method is described in our <a href="http://genomebiology.com/2014/15/10/501/abstract">paper</a> published in Genome Biology. The Bayesembler is being developed by Lasse Maretty, Jonas Andreas Sibbesen and Anders Krogh at the Bioinformatics Centre, a part of the <a href="http://www1.bio.ku.dk/binf/">Section for Computational and RNA Biology</a> at the <a href="http://www.ku.dk/english">University of Copenhagen</a>.</p>
      </section>
    </div>

    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer">
      <footer class="inner">
        <p class="copyright">Bayesembler maintained by <a href="https://github.com/bioinformatics-centre">bioinformatics-centre</a></p>
        <p>Published with <a href="http://pages.github.com">GitHub Pages</a></p>
      </footer>
    </div>

              <script type="text/javascript">
            var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
            document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
          </script>
          <script type="text/javascript">
            try {
              var pageTracker = _gat._getTracker("UA-44694489-1");
            pageTracker._trackPageview();
            } catch(err) {}
          </script>


  </body>
</html>
